webapp/src/ts/services/uhc-stats.service.ts

```diff
@@ -0,0 +1,105 @@
+import { Injectable } from '@angular/core';
+import * as moment from 'moment';
+import { isObject as _isObject, uniq as _uniq } from 'lodash-es';
+import * as CalendarInterval from '@medic/calendar-interval';
+
+import { DbService } from '@mm-services/db.service';
+import { ContactTypesService } from '@mm-services/contact-types.service';
+import { SessionService } from '@mm-services/session.service';
+import { AuthService } from '@mm-services/auth.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UHCStatsService {
+  hasHomeVisitPermission;
+
+  constructor(
+    private dbService: DbService,
+    private contactTypesService: ContactTypesService,
+    private sessionService: SessionService,
+    private authService: AuthService
+  ) { }
+
+  private async getLastVisitedDate(contactId) {
+    const records = await this.dbService
+      .get()
+      .query('medic-client/contacts_by_last_visited', { reduce: true, group: true, key: contactId });
+
+    const lastVisitedDateRow = records?.rows?.length ? records.rows[0] : {};
+
+    return _isObject(lastVisitedDateRow.value) ? lastVisitedDateRow.value.max : lastVisitedDateRow.value;
+  }
+
+  private async getVisitsInDateRange(dateRange: DateRange, contactId) {
+    const records = await this.dbService
+      .get()
+      .query(
+        'medic-client/visits_by_date',
+        { start_key: [ contactId, dateRange.start ], end_key: [ contactId, dateRange.end ] }
+      );
+
+    return records?.rows?.map(row => {
+      return moment(row.key[1])
+        .startOf('day')
+        .valueOf();
+    });
+  }
+
+  private async canViewHomeVisitStats() {
+    if (this.hasHomeVisitPermission !== undefined) {
+      return this.hasHomeVisitPermission;
+    }
+
+    const permission = 'can_view_uhc_stats';
+    // Disable UHC for DB admins.
+    this.hasHomeVisitPermission = this.sessionService.isDbAdmin() ? false : await this.authService.has(permission);
+
+    return this.hasHomeVisitPermission;
+  }
+
+  async getHomeVisitStats(contact, visitCountSettings: VisitCountSettings): Promise<VisitStats> {
+    if (!visitCountSettings || !contact) {
+      return;
+    }
+
+    const canView = await this.canViewHomeVisitStats();
+
+    if (!canView) {
+      return;
+    }
+
+    const typeId = this.contactTypesService.getTypeId(contact);
+    const type = await this.contactTypesService.get(typeId);
+
+    if (!type?.count_visits) {
+      return;
+    }
+
+    const lastVisitedDate = await this.getLastVisitedDate(contact._id);
+    const dateRange = CalendarInterval.getCurrent(visitCountSettings.monthStartDate);
+    const visits = lastVisitedDate >= dateRange?.start ? await this.getVisitsInDateRange(dateRange, contact._id) : [];
+
+    return {
+      lastVisitedDate: lastVisitedDate,
+      count: _uniq(visits).length,
+      countGoal: visitCountSettings.visitCountGoal
```

Comment:
We could do app services a favor and include the date range here! Right now they hardcode it in the config! 

We should probably make this available for all contacts, though, regardless of permissions settings and whatnot ? 
What do you think?

Reply:
> We could do app services a favor and include the date range here!

@ muso, I see they use it only for counting the visits. Do you think they might need it for something else? 
I think adding the date range would help explain better this object's values and perhaps it helps them debugging issues. 👍 

> We should probably make this available for all contacts

Do you mean removing the restrictions by contact.type and can_view_uhc_stats? Or just can_view_uhc_stats?
I think since we're now filtering the view's rows by contact id, it's less heavy for the users and it doesn't look like sensitive information yet, like to prevent other roles of viewing the stats.

Reply:
Same with DB Admin? 🤔

Reply:
I was just thinking that the settings (interval start and end) shouldn't be restricted based of anything (since they require almost no calculation, so has close to zero performance impact).

Reply:
What about if we leave the range out of the homeVisits scope, since maybe they can use it for other things outside home visits, for example:
```
const uhcStats = {
  homeVisits: await this.uhcStatsService.getHomeVisitStats(contact, this.visitCountSettings)
  uhcMonthRange: await this.uhcStatsService.getUHCMonthRange(this.visitCountSettings)   
};
```

Struggling with a proper name 😅

Reply:
> What about if we leave the range out of the homeVisits scope
that's great, and exactly what I meant :) 
> uhcMonthRange
`uhcInterval` ?

Reply:
Done!